README.md

```diff
@@ -1 +1,93 @@
-TODO.
+# Kubewarden policy volumemounts-policy
+
+## Description
+
+Policy that inspects containers, init containers, and ephemeral containers, and
+restricts their usage of volumes by  checking the `volume` name being used in
+`volumeMounts[*].name`.
+
+The policy can either target `Pods`, or [workload
+resources](https://kubernetes.io/docs/concepts/workloads/) (`Deployments`,
+`ReplicaSets`, `DaemonSets`, `ReplicationControllers`, `Jobs`, `CronJobs`) by
+setting the policy's `spec.rules` accordingly.
+
+Both have trade-offs:
+* Policy targets Pods: Different kind of resources (be them native or CRDs) can
+  create Pods. By having the policy target Pods, we guarantee that all the Pods
+  are going to be compliant, even those created from CRDs.
+  However, this could lead to confusion among users, as high level Kubernetes
+  resources would be successfully created, but they would stay in a non
+  reconciled state. Example: a Deployment creating a non-compliant Pod would be
+  created, but it would never have all its replicas running.
+* Policy targets workload resources (e.g: Deployment): the policy inspect higher
+  order resource (e.g. Deployment): users will get immediate feedback about
+  rejections.
+  However, non compliant pods created by another high level resource (be it
+  native to Kubernetes, or a CRD), may not get rejected.
+
+
+## Settings
+```yaml
+reject: anyIn # one of anyIn (default, denylist), anyNotIn (allowlist), allIn, allNotIn
+volumeMountNames:  # list of volumeMounts.name to match using the defined reject operator
+  - foo
+  - bar
+  - baz
+```
+
+- `allIn` checks if all of the volumeMountNames are in the Pod/Workload resource
+- `anyIn` checks if any of the volumeMountNames are in the Pod/Workload resource
+- `allNotIn` checks if all of the volumeMountNames are not in the Pod/Workload resource
+- `anyNotIn` checks if any of the volumeMountNames are not in the Pod/Workload resource
+
+## Examples
+
+```yaml
+# denylist, reject volumeMounts named `my-volume` or `my-volume2`
+reject: anyIn
+volumeMountsNames:
+  - my-volume
+  - my-volume2
+```
+
+```yaml
+# allowlist, only allow volumeMounts named `my-volume3` or `my-volume4`
+reject: anyNotIn
+volumeMounts:
+  - my-volume3
+  - my-volume4
+```
+
+```yaml
+# container cannot use both volumes at once, only one or the other
+reject: allIn
+volumeMounts:
+  - my-volume5
+  - my-volume6
+```
+
+```yaml
+# container can use both volumes at once, but not only one of them
+reject: allNotIn
+volumeMounts:
+  - my-volume5
+  - my-volume6
+```
```

Comment:
These are the two rules that I find hard to understand without reading the docs. I can get `allIn` with a **lot** of thinking.

However, I cannot get the `allNotIn`... to me it reads like:

> **reject** **all** the volume mount names that does not belong to this list `[my-volume5, my-volume6`]

which is basically like the `anyIn`...

Maybe it's my fault, I've been thinking about how we could clarify that by picking other key values, but I'm out of ideas

Reply:
I also find it hard to understand these two rules

`allIn` _# container cannot use both volumes at once, only one or the other_
_only one or the other_ does it mean it has to have one of the two volumes? is it ok if it doesn't have any of them?
if it should reject if it contains exactly these two rules I would name it `match` or `exactMatch` or something like that

`allNotIn` I see it the same way as Flavio:
> reject all the volume mount names that does not belong to this list [my-volume5, my-volume6]

Reply:
They are indeed super obtuse.
I went with those because they are closer to NV ones (`CriteriaOpContainsAll, CriteriaOpContainsAny, CriteriaOpNotContainsAny, CriteriaOpContainsOtherThan`), and also Kyverno ones (see [here](https://kyverno.io/docs/writing-policies/preconditions/#operators)).

The `reject` verb doesn't help, but it simplifies thinking about what's the end result of the policy. Because if not, one gets confused by matching the sets of names, and doesn't realize that matching the sets, means that the policy rejects.

`match` and `exactMatch` don't convey the meaning of "some but not all at once".

How about:
- `allIn` -> `notSimultaneouslyIn`
- `allIn` -> `AllAreUsed`. Reject container if all volume names are used.
- `allIn` -> `AllPresent`: Reject container if all are present

and:
- `allNotIn` -> `simultaneouslyIn`
- `allNotIn` -> `someButNotAllAreMissing`. Reject container if some (but not all) are missing.
- `allNotIn` -> `SomeMissing`: Reject container if some are missing


I think `AllPresent` and `SomeMissing` are the most easily understood.

Reply:
What about:

```yaml
# this replaces allIn
reject: allAreUsed
```

and

```yaml
# this replaces allNotIn
reject: someAreNotUsed
```

in this way "used" is adopted by these two criteria that are closely related

Reply:
Fine with me, go with `allAreUsed` and `notAllAreUsed` :+1:

Reply:
Done in https://github.com/kubewarden/volumeMounts-policy/pull/2/commits/8822a511699e279abd9628712651a3a749a9ca8b.

Reply:
`allAreUsed` and `notAllAreUsed` makes it more clear to me